cli/lsp/repl.rs

```diff
@@ -0,0 +1,292 @@
+  pub new_text: String,
+  pub span: Span,
+}
+
+pub struct ReplLanguageServer {
+  language_server: super::language_server::LanguageServer,
+  document_version: i32,
+  document_text: String,
+  pending_text: String,
+  cwd_uri: ModuleSpecifier,
+}
+
+impl ReplLanguageServer {
+  pub async fn new_initialized() -> Result<ReplLanguageServer, AnyError> {
+    super::logging::set_lsp_log_level(log::Level::Debug);
+    let language_server =
+      super::language_server::LanguageServer::new(Client::new_for_repl());
+
+    let cwd_uri = get_cwd_uri()?;
+
+    #[allow(deprecated)]
+    language_server
+      .initialize(InitializeParams {
+        process_id: None,
+        root_path: None,
+        root_uri: Some(cwd_uri.clone()),
+        initialization_options: Some(
+          serde_json::to_value(get_repl_workspace_settings()).unwrap(),
+        ),
+        capabilities: ClientCapabilities {
+          workspace: None,
+          text_document: None,
+          window: None,
+          general: None,
+          experimental: None,
+        },
+        trace: None,
+        workspace_folders: None,
+        client_info: Some(ClientInfo {
+          name: "Deno REPL".to_string(),
+          version: None,
+        }),
+        locale: None,
+      })
+      .await?;
+
+    language_server.initialized(InitializedParams {}).await;
+
+    let server = ReplLanguageServer {
+      language_server,
+      document_version: 0,
+      document_text: String::new(),
+      pending_text: String::new(),
+      cwd_uri,
+    };
+    server.open_current_document().await;
+
+    Ok(server)
+  }
+
+  pub async fn commit_text(&mut self, line_text: &str) {
+    self.did_change(line_text).await;
+    self.document_text.push_str(&self.pending_text);
+    self.pending_text = String::new();
+  }
+
+  pub async fn completions(
+    &mut self,
+    line_text: &str,
+    position: usize,
+  ) -> Vec<ReplCompletionItem> {
+    self.did_change(line_text).await;
+    let before_line_len = BytePos(self.document_text.len() as u32);
+    let position = before_line_len + BytePos(position as u32);
+    let text_info = deno_ast::SourceTextInfo::from_string(format!(
+      "{}{}",
+      self.document_text, self.pending_text
+    ));
+    let line_and_column = text_info.line_and_column_index(position);
+    let response = self
+      .language_server
+      .completion(CompletionParams {
+        text_document_position: TextDocumentPositionParams {
+          text_document: TextDocumentIdentifier {
+            uri: self.get_document_specifier(),
+          },
+          position: Position {
+            line: line_and_column.line_index as u32,
+            character: line_and_column.column_index as u32,
+          },
+        },
+        work_done_progress_params: WorkDoneProgressParams {
+          work_done_token: None,
+        },
+        partial_result_params: PartialResultParams {
+          partial_result_token: None,
+        },
+        context: None,
```

Comment:
discussion: while it isn't important import suggestions/completions, it is important for other types of completions, and it would be good to consider how to provide this information from the client.

Reply:
This is related to when to show completions, right? Right now completions only happen when hitting tab in the repl so I think we wouldn't need to configure this?

Reply:
[Then it would be](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#completionContext):

```rust
let context = CompletionContext {
  trigger_kind: CompletionTriggerKind::Invoked,
  trigger_character: None,
};
```